src/poseidon/traits.rs

```diff
@@ -0,0 +1,416 @@
+        let params_set = if !optimized_for_constraints {
+            P::PARAMS_OPT_FOR_CONSTRAINTS
+        } else {
+            P::PARAMS_OPT_FOR_WEIGHTS
+        };
+
+        for param in params_set.iter() {
+            if param[0] == rate {
+                let (ark, mds) = Self::find_ark_and_mds(
+                    P::MODULUS_BITS as u64,
+                    rate,
+                    param[2] as u64,
+                    param[3] as u64,
+                    param[4] as u64,
+                );
+
+                return Some(PoseidonParameters {
+                    full_rounds: param[2],
+                    partial_rounds: param[3],
+                    alpha: param[1] as u64,
+                    ark: ark,
+                    mds: mds,
+                    rate: param[0],
+                    capacity: 1,
+                });
+            }
+        }
+
+        None
+    }
+
+    /// Internal function that computes the ark and mds from the Poseidon Grain LFSR.
+    fn find_ark_and_mds(
+        prime_bits: u64,
+        rate: usize,
+        full_rounds: u64,
+        partial_rounds: u64,
+        skip_matrices: u64,
+    ) -> (Vec<Vec<Self>>, Vec<Vec<Self>>) {
+        let mut lfsr = PoseidonGrainLFSR::new(
+            false,
+            prime_bits,
+            (rate + 1) as u64,
+            full_rounds,
+            partial_rounds,
+        );
+
+        let mut ark = Vec::<Vec<Self>>::new();
+        for _ in 0..(full_rounds + partial_rounds) {
+            ark.push(lfsr.get_field_elements_rejection_sampling(rate + 1));
+        }
+
+        let mut mds = Vec::<Vec<Self>>::new();
+        mds.resize(rate + 1, vec![Self::zero(); rate + 1]);
+        for _ in 0..skip_matrices {
+            let _ = lfsr.get_field_elements_mod_p::<Self>(2 * (rate + 1));
+        }
+
+        // a qualifying matrix must satisfy the following requirements
+        // - there is no duplication among the elements in x or y
+        // - there is no i and j such that x[i] + y[j] = p
+        // - the resultant MDS passes all the three tests
+
+        let xs = lfsr.get_field_elements_mod_p::<Self>(rate + 1);
+        let ys = lfsr.get_field_elements_mod_p::<Self>(rate + 1);
+
+        for i in 0..(rate + 1) {
+            for j in 0..(rate + 1) {
+                mds[i][j] = (xs[i] + &ys[j]).inverse().unwrap();
+            }
+        }
+
+        (ark, mds)
+    }
+}
+
+macro_rules! impl_poseidon_default_parameters_field {
+    ($field: ident, $params: ident) => {
+        impl<P: $params + PoseidonDefaultParameters> PoseidonDefaultParametersField for $field<P> {
+            fn get_default_parameters(
+                rate: usize,
+                optimized_for_constraints: bool,
+            ) -> Option<PoseidonParameters<Self>> {
+                Self::get_default_parameters_internal::<P>(rate, optimized_for_constraints)
+            }
+        }
+    };
+}
+
+impl_poseidon_default_parameters_field!(Fp64, Fp64Parameters);
```

Comment:
Will we implement `PoseidonDefaultParameters` for `Fp...Parameters` in `ark_ff` or here?

Reply:
Nope, as the parameters would be field-specific, so later `ark-ec` will need to be added with additional information about the Poseidon parameters.

Reply:
Got it Thanks